### PR TITLE
AirfRANS: 3L/384d for volume improvement (deeper + wider)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

3L/384d provides more capacity than 3L/256d and stays within known stability bounds (2L/384d diverged, but 3L may be more stable). The extra width at 384d may help model the full-volume field better.

## Instructions

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --enable-fourier --model-layers 3 --model-hidden-dim 384 --model-heads 6 \
  --epochs 999 \
  --wandb_group af-volume-improvement
```

**Safety:** 2L/384d diverged (#2828: NaN from ep134). 3L may be more stable due to more gradient smoothing. If diverges before ep50, add `--grad-clip 0.5` and document.

## Reporting Requirements
- Best val `surface_mse` AND `volume_mse`
- Test pair at best val checkpoint
- W&B run ID

## Baseline

**AirfRANS baselines:**
- val surface_mse: `0.000482` @ ep576 (PR #2951, W&B: `pr4wsbfm`)
- Paper pair: Surf MSE = 0.003, Vol MSE = 0.00764 (from PR #2824)
- External target (SpiderSolver): Surf MSE 0.0043 / Vol MSE 0.0017
- **Surface is already below target; volume (0.00764 vs 0.0017) needs 4.5x improvement**

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT taki:

## Results

### 3L/384d AirfRANS — NEGATIVE RESULT (diverged)

| Metric | Best Value | Epoch | Baseline | Delta |
|--------|-----------|-------|----------|-------|
| val surface_mse | 0.01947 | 48 | 0.000482 | +3940% (40x worse) |
| val volume_mse | 0.08889 | 48 | 0.00764 (paper) | +1063% (11.6x worse) |
| test surface_mse | 0.02360 | 48 | — | — |
| test volume_mse | 0.10394 | 48 | — | — |
| Best val volume_mse | 0.08129 | 52 | 0.00764 (paper) | +964% (10.6x worse) |

**W&B run:** `kt3h6c9t` (taki/af-3l-384d-volume)
**Total epochs:** 380 (360-min budget)
**Best checkpoint:** epoch 48

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --enable-fourier --model-layers 3 --model-hidden-dim 384 --model-heads 6 \
  --epochs 999 --wandb-group af-volume-improvement \
  --agent taki --wandb-name "taki/af-3l-384d-volume"
```

### What happened

**Catastrophic gradient explosion at ep53.** The model showed promising early descent (0.377 → 0.0195 over 48 epochs), then gradient norms exploded from ~21 to 730 at the first T_max=50 cosine LR restart (ep50). The model never recovered — by ep380, val surface_mse had plateaued at 0.77.

Gradient norm timeline around divergence:
- ep48: grad_norm=18.4, val=0.0195 (best)
- ep52: grad_norm=21.8, val=0.0205
- ep53: grad_norm=40.8, val=0.0827 (spike)
- ep54: grad_norm=730.5, val=0.2953 (catastrophic)
- ep55+: never recovered

**Key insight:** 3L/384d is too wide for stable AirfRANS training with gc=1.0. The divergence pattern is similar to 2L/384d (PR #2828, NaN at ep134) but hits earlier (ep53 vs ep134), likely because the cosine T_max=50 restarts are more aggressive. Even at its best checkpoint (ep48), the 3L/384d model (0.0195) is 40x worse than the 2L/256d champion (0.000482 at ep576). The 3L depth does NOT compensate for 384d width instability — contrary to the hypothesis that more gradient smoothing from additional layers would stabilize wider models.

**Safety note result:** The PR predicted divergence risk and suggested gc=0.5 fallback if divergence before ep50. Divergence occurred at ep53 (borderline). Given the model was already 40x off baseline even at its best, gc=0.5 is unlikely to close the gap.

### Suggested follow-ups
- The AirfRANS 2L/256d architecture appears to be a hard constraint — width beyond 256d diverges regardless of depth (2L/384d, 3L/384d both failed)
- If more capacity is needed for volume improvement, try **2L/320d** as a safer intermediate step, or **3L/256d** (different depth, same width) which reached 0.001241 (PR #2820)
- Volume improvement may be better pursued through loss weighting, multi-scale features, or volume-specific augmentation rather than raw model capacity

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3100 because ADVISOR: Catastrophic divergence at ep53 (first cosine restart) — surface=0.01947/vol=0.08889 (>30x worse than baseline). This matches the earlier 2L/384d failure (NaN). The 2L/256d architecture appears to be a hard stability constraint for AirfRANS under the current LR/cosine regime. Wider/deeper architectures consistently destabilize. Closing — architecture direction dead for AirfRANS..
